pkg/querier/api/DESIGN.md

```diff
@@ -0,0 +1,81 @@
+| --------------------------------------------------------------------------------------------------------- | ----------- | ----------- | ------------ |
+| Comma-separated string: "<partition id>:<offset id>"                                                      | N/A         | N/A         | N/A          |
+| Binary-encoded list of <partition id> <offset id> pairs, encoded with: Varint + base64                    | 64179 ns/op | 114722 B/op | 49 allocs/op |
+| Binary-encoded list of <partition id> <offset id> pairs, encoded with: Varint + snappy + base64           | 58544 ns/op | 105252 B/op | 48 allocs/op |
+| Binary-encoded list of <offset id> only, encoded with: Varint + snappy + base64                           | 46693 ns/op | 90072 B/op  | 31 allocs/op |
+| Binary-encoded list of <partition id> <offset id> pairs, encoded with: Int32 / int64 + base64             | 56468 ns/op | 114724 B/op | 49 allocs/op |
+| Binary-encoded list of <partition id> <offset id> pairs, encoded with: Int32 / int64 + snappy + base64    | 54670 ns/op | 114724 B/op | 49 allocs/op |
+| Binary-encoded list of <offset id> only, encoded with: Int + snappy + base64                              | 40447 ns/op | 88027 B/op  | 31 allocs/op |
+
+### Lookup performance
+
+The lookup is the process to read 1 specific partition offset from the HTTP header value string.
+
+| Encoding format                                                                                           | CPU         | Memory      | Allocations  |
+| --------------------------------------------------------------------------------------------------------- | ----------- | ----------- | ------------ |
+| Comma-separated string: "<partition id>:<offset id>"                                                      | 4200 ns/op  | 3 B/op      | 1 allocs/op  |
+| Binary-encoded list of <partition id> <offset id> pairs, encoded with: Varint + base64                    | 23399 ns/op | 48193 B/op  | 13 allocs/op |
+| Binary-encoded list of <partition id> <offset id> pairs, encoded with: Varint + snappy + base64           | 18063 ns/op | 38720 B/op  | 12 allocs/op |
+| Binary-encoded list of <offset id> only, encoded with: Varint + snappy + base64                           | 13120 ns/op | 24384 B/op  | 10 allocs/op |
+| Binary-encoded list of <partition id> <offset id> pairs, encoded with: Int32 / int64 + base64             | 18974 ns/op | 48192 B/op  | 13 allocs/op |
+| Binary-encoded list of <partition id> <offset id> pairs, encoded with: Int32 / int64 + snappy + base64    | 17451 ns/op | 48193 B/op  | 13 allocs/op |
+| Binary-encoded list of <offset id> only, encoded with: Int + snappy + base64                              | 9615 ns/op  | 22336 B/op  | 10 allocs/op |
+
+### Considerations
+
+- The Mimir use case is encode once (in the query-frontend) and decode / lookup many (in each ingester)
```

Comment:
I see it as an optimization not a design thing. It's something we can consider for the future, but I wouldn't block on this.

Morever, I want to stress that this PR is just the first step. I don't want to wait even 1ns in the ingesters as final goal. My idea is that medium terms these offsets won't be propagated at all, and all the waiting will happen in the query-frontend. However, since we haven't designed the 2nd step of the work yet, I can't say that it will happen for sure, but that's my goal.